beets/dbcore/db.py

```diff
@@ -1059,47 +1085,94 @@ def _make_attribute_table(self, flex_table):
 
     # Querying.
 
+    @staticmethod
+    def _get_fields(query):
+        if hasattr(query, "subqueries"):
+            return reduce(add, map(Database._get_fields, query.subqueries))
+        if hasattr(query, "subquery"):
+            return Database._get_fields(query.subquery)
+        elif hasattr(query, "fields"):
+            return [(f, True) for f in query.fields]
+        elif hasattr(query, "field"):
+            return [(query.field, query.fast), ]
+        else:
+            return []
+
     def _fetch(self, model_cls, query=None, sort=None, limit=None):
         """Fetch the objects of type `model_cls` matching the given
         query. The query may be given as a string, string sequence, a
         Query object, or None (to fetch everything). `sort` is an
-        `Sort` object.
+        `Sort` object, `limit` is an integer or None.
         """
-        query = query or TrueQuery()  # A null query.
-        sort = sort or NullSort()  # Unsorted.
-        where, subvals = query.clause()
+        sort = sort or NullSort()
         order_by = sort.order_clause()
 
-        sql = ("SELECT * FROM {} WHERE {} {} {}").format(
-            model_cls._table,
-            where or '1',
-            f"ORDER BY {order_by}" if order_by else '',
-            f"LIMIT {limit}" if limit else '',
-        )
+        query = query or TrueQuery()
+        where, subvals = query.clause()
 
-        # Fetch flexible attributes for items matching the main query.
-        # Doing the per-item filtering in python is faster than issuing
-        # one query per item to sqlite.
-        flex_sql = ("""
-            SELECT * FROM {} WHERE entity_id IN
-                (SELECT id FROM {} WHERE {});
-            """.format(
-            model_cls._flex_table,
-            model_cls._table,
-            where or '1',
-        )
-        )
+        table = model_cls._table
+
+        fields = {}
+        for fast, _fields in groupby(self._get_fields(query), lambda x: x[1]):
+            fields[fast] = {f[0] for f in _fields}
+
+        flex_fields = fields.get(False, set())
+        other_model_fields = fields.get(True, set()) - set(model_cls._fields)
```

Comment:
`groupby` expects its input to be sorted: https://docs.python.org/3/library/itertools.html#itertools.groupby. From what I can tell, this is not the case here.

One fix would be to drop the `groupby` entirely:
```python
fields = defaultdict(set)
for field, fast in self._get_fields(query):
    fields[fast].add(field)

flex_fields, model_fields = fields[False], fields[True]
other_model_fields = model_fields - set(model_cls._fields)
```

Reply:
Good point!